daemon/containerd/image_exporter.go

```diff
@@ -1,19 +1,66 @@
 package containerd
 
-import "io"
+import (
+	"context"
+	"io"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/images/archive"
+	"github.com/containerd/containerd/platforms"
+	"github.com/docker/distribution/reference"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
 
 // ExportImage exports a list of images to the given output stream. The
 // exported images are archived into a tar when written to the output
 // stream. All images with the given tag and all versions containing
 // the same tag are exported. names is the set of tags to export, and
 // outStream is the writer which the images are written to.
-func (i *ImageService) ExportImage(names []string, outStream io.Writer) error {
-	panic("not implemented")
+func (i *ImageService) ExportImage(ctx context.Context, names []string, outStream io.Writer) error {
+	opts := []archive.ExportOpt{
+		archive.WithPlatform(platforms.Ordered(platforms.DefaultSpec())),
+		archive.WithSkipNonDistributableBlobs(),
+	}
+	is := i.client.ImageService()
+	for _, imageRef := range names {
+		named, err := reference.ParseDockerRef(imageRef)
+		if err != nil {
+			return err
+		}
+		opts = append(opts, archive.WithImage(is, named.String()))
+	}
+	return i.client.Export(ctx, outStream, opts...)
 }
 
 // LoadImage uploads a set of images into the repository. This is the
 // complement of ExportImage.  The input stream is an uncompressed tar
 // ball containing images and metadata.
-func (i *ImageService) LoadImage(inTar io.ReadCloser, outStream io.Writer, quiet bool) error {
-	panic("not implemented")
+func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, outStream io.Writer, quiet bool) error {
+	platform := platforms.DefaultStrict()
+	imgs, err := i.client.Import(ctx, inTar, containerd.WithImportPlatform(platform))
+
+	if err != nil {
+		logrus.WithError(err).Error("Failed to import image to containerd")
+		return errors.Wrapf(err, "Failed to import image")
```

Comment:
Looks like we're both _logging_ and returning the error here; I'd have to check, but I suspect that means we're logging the error _twice_.

(I should also change the `errors.Wrapf` to `errors.Wrap` and `Failed -> failed`)

Reply:
OK; had a quick try, and I was incorrectly assuming that we were logging errors for API responses, so it's not logged twice.
 
That said; I think there's a couple of things we should look into (some of that "in general")

- Perhaps we should consider logging internal server errors (5xx) when they occur (there's little use for logging invalid requests (4xx), although we could consider logging those as "debug").
- In debug mode, we do log errors if they have no `errdefs` type associated, as part of [`api/server/httpstatus.FromError()`](https://github.com/moby/moby/blob/301eba7fa4871862fbc00753c4accfc535e0eedf/api/server/httpstatus/status.go#L68-L71), which is used in [`api/server.makeErrorHandler()`](https://github.com/moby/moby/blob/301eba7fa4871862fbc00753c4accfc535e0eedf/api/server/errorhandler.go#L16-L18) (and some other places).
- _HOWEVER_ for endpoints that return a stream, we don't (can't!) return a 4xx or 5xx status because headers have already been sent at that point, so the error is [included in the JSON stream](https://github.com/moby/moby/blob/301eba7fa4871862fbc00753c4accfc535e0eedf/api/server/router/image/image_routes.go#L166-L168) and we may have to update [`pkg/streamformatter.FormatError()`](https://github.com/moby/moby/blob/301eba7fa4871862fbc00753c4accfc535e0eedf/pkg/streamformatter/streamformatter.go#L31-L41) to allow passing a logger.
 
 
For this PR;

- I expect most errors would be caused by invalid requests (at least for _load_), however, a quick glance at the containerd code showed that many errors may not be "typed" (no containerd/errdefs type), so it may be difficult to distinguish internal errors from invalid requests.
- Given that this is all new code, and that there may still be code-paths we could hit and bugs to fix, I think we can keep _some_ of these logs, but change them to be a `Warn()` (at most)
- I'll add suitable TODO's so that we can revisit these later (either "remove" the logs, or change them to a "debug")

Reply:
I created https://github.com/moby/moby/issues/43909 for the "in general" tracking.